Add letsencrypt support for the repository host.

nuclearsandwich opened the pull request:
Like the Jenkins host, the repo host may also be publicly exposed. Apt repositories don't require https for repository integrity since that's achieved through GPG key signing repository metadata.

However other resources on the repository host may benefit from having a secure transport mechanism. Including the secure distribution of the public key information used for the apt security process or other artifacts hosted on the repository server.

nuclearsandwich committed:
Add ros_buildfarm.repo.server_name attribute.

I am bit conflicted about defaulting this to `nil`, the jenkins server
name defaults to `'ros_buildfarm'` and would biff if used with
letsencrypt enabled since it isn't a valid fqdn for the server name
verification.

`nil` breaks the convention of having a default but also gives us the
ability to raise the alarm early if a configuration isn't going to work,
rather than having the acme.sh certificate issue attempt fail.

nuclearsandwich committed:
Extract acme.sh installation into a recipe.

Alternatively, we could make it a resource housed in the cookbook but
this is fine for the time being.

nuclearsandwich committed:
Add letsencrypt support for repo host.

Like the Jenkins host, the repo host may also be publicly exposed.
Apt repositories don't require https for repository integrity since
that's achieved through GPG key signing repository metadata.

However other resources on the repository host may benefit from having a
secure transport mechanism. Including the secure distribution of the
public key information used for the apt security process or other
artifacts hosted on the repository server.

nuclearsandwich commented on attributes/repo.rb:
````diff
@@ -21,6 +21,10 @@
 # ```
 default['ros_buildfarm']['repo']['rsyncd_endpoints'] = Hash[]
 
+# The repository host server name is currently only used to provide
+# a server name when using letsencrypt for https and is environment-specific.
+default['ros_buildfarm']['repo']['server_name'] = nil
````
I am bit conflicted about defaulting this to `nil`, the jenkins server
name defaults to `'ros_buildfarm'` and would biff if used with
letsencrypt enabled since it isn't a valid fqdn for the server name
verification.

`nil` breaks the convention of having a default but also gives us the
ability to raise the alarm early if a configuration isn't going to work,
rather than having the acme.sh certificate issue attempt fail.

nuclearsandwich commented on recipes/repo.rb:
```diff
@@ -689,12 +689,53 @@
+    server_name: server_name,
   ]
-  notifies :restart, 'service[nginx]'
+  notifies :restart, 'service[nginx]', :immediately
```
We need this configuration to be applied before the acme certificate issuance is triggered below.

nuclearsandwich commented on templates/nginx/repo.conf.erb:
```diff
@@ -1,8 +1,20 @@
 server {
 	listen 80 default_server;
-	server_name repo;
+	<% if @letsencrypt_enabled %>
+	listen 443 ssl default_server;
```
`default_server` here is a per-address and per-port setting so these two do not conflict with each other.

The directive makes it so that even if someone is just requesting the server by IP address this site will be served for HTTP (80) or HTTPS (443) even without knowing the requested server name.

nuclearsandwich commented on recipes/repo.rb:
```diff
@@ -689,12 +689,53 @@
+
 template '/etc/nginx/sites-available/repo' do
   source 'nginx/repo.conf.erb'
   variables Hash[
-    rpm_repos: node['ros_buildfarm']['rpm_repos']
+    letsencrypt_enabled: node['ros_buildfarm']['letsencrypt_enabled'],
+    cert_path: cert_path,
+    key_path: key_path,
+    rpm_repos: node['ros_buildfarm']['rpm_repos'],
+    server_name: server_name,
   ]
-  notifies :restart, 'service[nginx]'
+  notifies :restart, 'service[nginx]', :immediately
 end
```
Unlike the Jenkins configuration, we make all the same content available via HTTP or HTTPS instead of redirecting to HTTPS.

Thus it makes more sense to use a single template with some variable options than it does to use separate templates as we do for the Jenkins host nginx config.

nuclearsandwich commented:
The yamllint check appears to be going through some strife upstream at the moment. I'm content to ignore it for a little until it settles down. I'm more perturbed by the Jenkins agent failure which looks like it could be a configuration race with authentication. I'll investigate that for sure.

cottsay reviewed (approved):
Thanks for the review guides. This looks great.

nuclearsandwich committed:
As of acme 3.0 letsencrypt is no longer the default but is still selectable.

nuclearsandwich commented:
I can also reproduce the jenkins cookbook CI failure on `latest`. So I'm going to make a case that we override the failed check and merge this since the recipe that we primarily changed is passing.

That being said, I _did_ move some stuff around in the jenkins recipe so it's possible that I've introduced a regression hiding behind the current failure but I'm willing to assume that risk and address it in a follow-up PR after I've fixed the current issue on another branch.
